src/tango/quic/fd_quic.c

```diff
@@ -1665,15 +1809,81 @@ fd_quic_handle_v1_handshake(
   return tot_sz;
 }
 
-ulong
-fd_quic_handle_v1_retry( fd_quic_t * quic, fd_quic_conn_t * conn, fd_quic_pkt_t const * pkt, uchar const * cur_ptr, ulong cur_sz ) {
+ulong fd_quic_handle_v1_retry(
+    fd_quic_t *           quic,
+    fd_quic_conn_t *      conn,
+    fd_quic_pkt_t const * pkt,
+    uchar const *         cur_ptr,
+    ulong                 cur_sz
+) {
   (void)pkt;
-  (void)quic;
-  (void)conn;
-  (void)cur_ptr;
-  (void)cur_sz;
-  DEBUG( FD_LOG_DEBUG(( "stub" )); )
-  return 0;
+  fd_quic_retry_t retry_pkt;
+  fd_quic_decode_retry(&retry_pkt, cur_ptr, cur_sz);
+
+  fd_quic_conn_id_t * orig_dst_conn_id = &conn->peer->conn_id;
+
+  /* Validate the Retry Integrity Tag. TODO can we make this more efficient? */
+  fd_quic_retry_pseudo_t retry_pseudo_pkt = {
+      .odcid_len        = orig_dst_conn_id->sz,
+      // .odcid
+      .hdr_form         = retry_pkt.hdr_form,
+      .fixed_bit        = retry_pkt.fixed_bit,
+      .long_packet_type = retry_pkt.long_packet_type,
+      .unused           = retry_pkt.unused,
+      .version          = retry_pkt.version,
+      .dst_conn_id_len  = retry_pkt.dst_conn_id_len,
+      // .dst_conn_id
+      .src_conn_id_len  = retry_pkt.src_conn_id_len,
+      // .src_conn_id
+      // .retry_token
+  };
+
+  /* Retry-pseudo packet is the retry packet with ODCID prepended and integrity
+     tag removed*/
+  memcpy( retry_pseudo_pkt.odcid, orig_dst_conn_id->conn_id, orig_dst_conn_id->sz );
+  memcpy( retry_pseudo_pkt.dst_conn_id, retry_pkt.dst_conn_id, retry_pkt.dst_conn_id_len );
+  memcpy( retry_pseudo_pkt.src_conn_id, retry_pkt.src_conn_id, retry_pkt.src_conn_id_len );
+  memcpy( retry_pseudo_pkt.retry_token, retry_pkt.retry_token, FD_QUIC_RETRY_TOKEN_SZ );
+
+  ulong retry_pseudo_footprint = fd_quic_encode_footprint_retry_pseudo( &retry_pseudo_pkt );
+  uchar retry_pseudo_buf[retry_pseudo_footprint];
+  fd_quic_encode_retry_pseudo( retry_pseudo_buf, retry_pseudo_footprint, &retry_pseudo_pkt );
+
+  /* Validate the retry integrity tag */
+  int rc = fd_quic_retry_integrity_tag_decrypt(
+      retry_pseudo_buf, (int)retry_pseudo_footprint, retry_pkt.retry_integrity_tag
+  );
+
+  /* Clients MUST discard Retry packets that have a Retry Integrity Tag that
+     cannot be validated */
```

Comment:
it functions more like a checksum and address validation of the server for the client rather than getting routed to the wrong endpoint. added a comment from RFC